### PR TITLE
Adds Initial Service to Get Posts

### PR DESCRIPTION
**Pull request by HyperTHD**
## Issue This PR Addresses
This PR address #1735 

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [x] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description

This PR is a work-in-progress, meant to work on adding a Post Microservice. This microservice parses posts from user's feeds and returns them. There's a path that returns the 10 latest posts, and you can also retrieve information about a specific post using it's ID.

This PR includes files that are being used by other microservices. I recommend getting those PRs in first before this one so that I can update the docker-compose file to use those services. This would allow me to remove the files those services used and only keep what I need for my microservice. In particular, this would be the User Microservice, The Post parsing microservice (#1828), and possibly the Feed Parsing microservice (#1833).

UPDATE: Screenshot of successful loading of this route along with network tab open as proof
Posts-Microservice

## Checklist

- [ ] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

**Comment by HyperTHD** [comment]
This PR is not tested btw. I'm still working on the Docker related stuff. I will post screenshots of the output when I get it working

File: src/api/posts/server.js
Comment (humphd) [inline src/api/posts/server.js:6]: You can do this at the root `/` vs. `/posts` since Traefik will add a `/v1/posts` prefix for you.

File: src/api/posts/src/bin/article-error.js
Comment (humphd) [inline src/api/posts/src/bin/article-error.js:1]: Let's get rid of the `bin` directory completely, everything can live in src.

File: src/api/posts/src/routes/posts.js
Comment (humphd) [inline src/api/posts/src/routes/posts.js:111]: Let's add a some more routes:

- `posts.post('/:id', ...)` for adding
- `posts.delete('/:id', ...)` for deleting.  Not sure if we need this one, or if we'd just delete at the feed level.  It's hard to remove a post, since it will just pop-up again the next time we crawl the feeds.
- `posts.put('/:id', ...)` for updating

All of these need to use the JWT protection I've added in Satellite (e.g., you have to be authenticated).  Furthermore, we'll need to deal with a user vs. admin:

- admin can do the operations above on any post (i.e., they don't need to own it)
- a user will only be able to do it for a post they own (i.e., belonging to a feed they own)

File: src/api/readme.md
Comment (humphd) [inline src/api/readme.md:41]: Why this change?

**Review by humphd — CHANGES_REQUESTED** [review]
I did a first read through and left some feedback.

There's a lot of code in here that can be removed and done in #1828 by @c3ho (e.g., all the feed handling stuff).

Eventually I want to have the feed service use this service to get/set/update/delete posts.

File: src/api/readme.md
Comment (HyperTHD) [reply src/api/readme.md:41]: Yikes, it seems when I updated the README and saved, it added those angle brackets together. That's something with the prettier settings iirc. I'll have to check on this

**Comment by HyperTHD** [comment]
UPDATE: 

- I added the docker code to create the posts image. I may have done it wrong, but I needed to include redis and elasticsearch inside the docker-compose.yml file since it needs both to function.
- I added the redis and elasticsearch information to all 3 env example files for reason specified above.
- I modified the redis and elasticsearch host names so that they use the container names. Since we're running them off the container instead of localhost, it makes the most sense to do this.
- Lastly, I refactored the code to remove the bin folder completely.

Running the posts service atm loads the right page, however, NO POSTS ARE LOCATED AND YOU GET AN EMPTY ARRAY AS A RESULT. I feel this is because of the changed host names, since the redis data is locally served instead of via the cloud. I'm not too sure about this one. Otherwise, I feel good about this service and can move to add in the extra routes

File: src/api/posts/src/bin/article-error.js
Comment (HyperTHD) [reply src/api/posts/src/bin/article-error.js]: Done

File: src/api/posts/server.js
Comment (HyperTHD) [reply src/api/posts/server.js]: Didn't know about that, nice to know, thanks!

File: src/api/docker-compose.yml
Comment (humphd) [inline src/api/docker-compose.yml:53]: We don't want to expose Redis to the big-bad world.  We made this mistake once before, and got hacked, see https://docs.google.com/presentation/d/1xBYQyNEDJPS1JRlzw4JHAX1exuPTmTB-a950zHiU950/.  In Docker Compose, defining ports works like this:

```
ports:
  # Port definitions can take 1 or 2 values:
  # DOCKER_HOST_PORT: DOCKER_CONTAINER PORT
  6379: 6379
  # The above means: "use port 6379 on the HOST machine to expose port 6739 in the CONTAINER"
  # It has the effect of connecting the Redis server's port running in the container to the same port in
  # the host computer, effectively opening that port through to the internet, if the host is running on
  # the web (staging, production). 
```

This is what people tend to do when they work on development, and need to connect directly to some internal database or server.  By connecting the container's port to your local dev machine's port, you can access things running in the container.  It works great.  But in a production environment, you don't want to expose traffic via ports like this.  We want to route everything through Traefik, which can manage load-balancing, authentication requests, and container/port mappings.

For access to Redis and Elastic from the outside world, we want to use your service as our only direct means.  It shouldn't be possible to connect to https://telescope.cdot.systems:6739 for example.  Your service (and others that will talk to Redis/Elastic) will only be available on the outside via a set of well-defined HTTP endpoints, which can use JWT based authentication (see #1796).

The right way to do what you're doing here is to do something like you can see in these PRs:

1. https://github.com/Seneca-CDOT/telescope/pull/1840/files#diff-a3872c2e4d094f394e87eb51d5c934953e79319accea61231188218ba9b184b0  In this example I'm running a container for the Firebase emulator.  It uses all kinds of ports, but I only want to enable 2 of them: 4000 (Firestore) and 8088 (the Dashboard UI).  Notice that I say `- '4000'` (i.e., only 1 port).  I do this because I am defining the port that's being used in the container, but I am not exposing it to the outside world.  Instead, I do this to let Traefik map that internal port to an endpoint URL.  Going to http://ui.firebase.localhost will allow a developer to get to the container's port 4000, but I don't open port 4000 on the host machine:

```
      - 'traefik.http.routers.firebase_ui.rule=Host(`ui.firebase.localhost`)'
      - 'traefik.http.routers.firebase_ui.service=firebase_ui'
      - 'traefik.http.services.firebase_ui.loadbalancer.server.port=4000'
```

In the case of `8088` I do both a container port and a host port mapping.  I don't want to do this, but the Firebase UI expects to be able to use localhost:8088 in order to load JS/CSS and other static assets in the browser.  Because this is a development situation, I'm not that worried.  But FYI.

2. You can see what @raygervais is doing in https://github.com/Seneca-CDOT/telescope/pull/1802/files#diff-9b5638ed60397f4c9b63cc41ff257015946394e745dce431fda440513146f5de with the Search service and Elastic.  Here he defines the elastic port on the container, but doesn't open it to the host.

The docker-compose files are evolving quickly in various PRs that are in review right now, so don't worry too much about making this perfect.  I would advise that the easiest fix here is to drop the second port from each of these services so we don't expose anything.

I wanted to mention it to you so you'd know, since this isn't an obvious thing, and the cost of getting it wrong is much larger than you realize.

File: src/api/docker-compose.yml
Comment (humphd) [inline src/api/docker-compose.yml:56]: I wonder where we should put volumes related to the microservices.  The root is one option for sure, but maybe we should have a dedicated directory where we can find these things.  cc @manekenpix, @raygervais.

File: src/api/docker-compose.yml
Comment (humphd) [inline src/api/docker-compose.yml:65]: Same thing here.

File: src/api/posts/src/processHTML/dom.js
Comment (humphd) [inline src/api/posts/src/processHTML/dom.js:1]: We won't do any of the HTML parsing/processing of feeds in this service, you can remove this dir.

**Review by humphd — CHANGES_REQUESTED** [review]

File: src/api/posts/src/processHTML/dom.js
Comment (HyperTHD) [reply src/api/posts/src/processHTML/dom.js:1]: Removing it right now will prevent my service from running at all until this process is merged. I'd either need to wait until the service is merged or just merge it there and have the person writing that service remove the directory and modify this service to use theirs

File: src/api/docker-compose.yml
Comment (manekenpix) [reply src/api/docker-compose.yml:56]: @humphd this is how we currently do it on prod and dev:

[Directory Structure](https://github.com/Seneca-CDOT/telescope/blob/master/docs/staging-production-deployment.md#directory-structure)

File: src/api/docker-compose.yml
Comment (HyperTHD) [reply src/api/docker-compose.yml:53]: Interesting, I forgot that the second port being there exposes the service to the web via the host machine. Noted

File: src/api/posts/src/routes/posts.js
Comment (HyperTHD) [reply src/api/posts/src/routes/posts.js:111]: Filed #1865 to handle these additional routes. This will likely involve talking with the User microservice or the Feed parsing microservice. Needs more discussion atm before we continue with this. Using the JWT protection in satellite should be easy enough

**Comment by manekenpix** [comment]
I need some clarification for this PR. I thought this microservice was going to answer requests (serving posts), but I see a lot of code here related to processing feeds and posts. Shouldn't all that be part of a different microservice?

@HyperTHD @humphd

**Comment by humphd** [comment]
@manekenpix I left a similar review comment, a lot of this code can get deleted.

**Comment by HyperTHD** [comment]
@manekenpix @humphd The main problem with removing that code is that there's no way for me to really know if the service works or not. The files related to my service also require the other files needed for feeds as well. Since we don't have those other services up, I can't realistically remove those extra files without breaking CI

**Comment by raygervais** [comment]
Why not test using mock services to pass in hypothetical data @HyperTHD ?

**Comment by HyperTHD** [comment]
@raygervais I plan on doing that once I remove the unnecessary files here. The thing I'm having here is that I need to figure out what to remove. I see that I don't need elasticsearch so I'm gonna remove all relevant files for that. Since this service just makes the requests to redis and fetches the posts data, I only need files relevant to that, including Feeds possibly, I'm not sure on that.

**Comment by HyperTHD** [comment]
@humphd Idk if Feeds or anything related to that is needed, but I kept that since the identity of a Post is based on that. Based on my testing, I realized that I'm only making requests to redis, so I removed all elasticsearch related files and other files that weren't needed. Unless I need to remove the feed stuff, this should be good and I can proceed with writing the tests

**Comment by c3ho** [comment]
I thought we were using firebase to store feeds and elastic to store posts? In that case wouldn't redis only be used for bull-queue itself to manage queues?

**Comment by humphd** [comment]
If/when we add a `POST` route to this, we'll need Elastic.  When a post gets added to the system, we do 2 things: cache it in Redis; index it in Elastic.

For the `GET` requests, we don't need Elastic (that's more for Search, see #1737).

The tension here around "how do I know if my code is working" is really best addressed through tests, which all services need.  The way you'd test this is to have the test harness put some posts into Redis (or our mock redis) and then try to get them again via `GET` requests. That's the best way to proceed.

File: src/api/docker-compose.yml
Comment (humphd) [inline src/api/docker-compose.yml:109]: This looks wrong.  Can you rebase?  I've updated how the docker-compose files are structured.

File: src/api/docker-compose.yml
Comment (humphd) [inline src/api/docker-compose.yml:121]: I wonder if we should move this volume into `src/api`?

File: src/api/docker-compose.yml
Comment (humphd) [inline src/api/docker-compose.yml:123]: If you want to remove this now, and let @raygervais add it in #1737, that's fine.

File: src/api/docker-compose.yml
Comment (humphd) [inline src/api/docker-compose.yml:155]: I need to research this more, but I don't think you can reuse this name `strip_api_prefix` again, since it's been used on the image service.  Probably it needs to be `strip_image_prefix` and `strip_posts_prefix` etc.

File: src/api/env.development
Comment (humphd) [inline src/api/env.development:18]: Let's simplify the code and just pass the URL in, removing port and password, since we don't mess with either in Docker.

**Review by humphd — DISMISSED** [review]

File: src/api/docker-compose.yml
Comment (HyperTHD) [reply src/api/docker-compose.yml:121]: I can move it there and remove it from the root folder.

Comment (humphd) [reply src/api/docker-compose.yml:121]: Don't remove that one, it's our legacy one that's being used by the current setup.  I'm just thinking about isolating these two from each other while we get the services working.

File: src/api/docker-compose.yml
Comment (HyperTHD) [reply src/api/docker-compose.yml:155]: This might have been the root cause of the traefik http routing issue I had whenever I would run both image and posts services together. I'll test and resolve if this works

File: src/api/docker-compose.yml
Comment (HyperTHD) [reply src/api/docker-compose.yml:109]: Yeah, this was the result of the merge conflicts resolution thing I was fixing. Will update

File: src/api/posts/package.json
Comment (humphd) [inline src/api/posts/package.json:7]: We should add a couple of scripts so you can run this locally (start, dev).  See image service.

File: src/api/posts/package.json
Comment (humphd) [inline src/api/posts/package.json:21]: Not necessary, you can remove.

File: src/api/posts/package.json
Comment (humphd) [inline src/api/posts/package.json:26]: Change this to `"^1.x"`

File: src/api/posts/package.json
Comment (humphd) [inline src/api/posts/package.json:27]: Crypto is built-into node.js now, you shouldn't need this.  Also, if the only reason you're pulling this in is to do the hash function, you can use what @Metropass added to Satellite:

```js
const { hash } = require('@senecacdot/satellite');
```

File: src/api/posts/src/data/post.js
Comment (humphd) [inline src/api/posts/src/data/post.js:3]:
```js
const { hash } = require('@senecacdot/satellite');
```

File: src/api/posts/src/storage.js
Comment (humphd) [inline src/api/posts/src/storage.js:28]: In a follow-up PR, we can rip out the feed code from this service, and/or do it when we bring in the feed queue work (cc @c3ho).

File: src/api/posts/test/posts.test.js
Comment (humphd) [inline src/api/posts/test/posts.test.js:37]:
```
    expect(Array.isArray(res.body)).toBe(true);
```

File: src/api/posts/test/posts.test.js
Comment (humphd) [inline src/api/posts/test/posts.test.js:38]: The way you're writing these tests, you probably don't need `done()` here, it's more when you have a callback function (e.g., if you used supertest's `end(function(...){...})` or something).  It's OK to leave it, or you can try removing too.

File: src/api/posts/test/posts.test.js
Comment (humphd) [inline src/api/posts/test/posts.test.js:45]: Instead of checking this in every test, write one test that does it and you can ignore it in all the others.

File: src/api/posts/test/posts.test.js
Comment (humphd) [inline src/api/posts/test/posts.test.js:46]: Same thing here and below, you're duplicating tests.

File: src/api/posts/test/posts.test.js
Comment (humphd) [inline src/api/posts/test/posts.test.js:97]: This doesn't look like it matches the code it goes with.

File: src/api/posts/test/posts.test.js
Comment (humphd) [inline src/api/posts/test/posts.test.js:102]: Write another test like this, but use the `url` you get vs. the `id`, to make sure both work.

File: src/api/posts/test/posts.test.js
Comment (humphd) [inline src/api/posts/test/posts.test.js:114]: These two checks aren't necessary (we just want a 404)

**Review by humphd — CHANGES_REQUESTED** [review]

**Comment by humphd** [comment]
I filed #1907 and #1908 for follow-up.

File: src/api/posts/Dockerfile
Comment (humphd) [inline src/api/posts/Dockerfile:1]: Typo here: `alpine` vs. `alphine`

**Review by humphd — CHANGES_REQUESTED** [review]

File: src/api/posts/src/storage.js
Comment (HyperTHD) [reply src/api/posts/src/storage.js:28]: Sounds good

File: src/api/docker-compose.yml
Comment (humphd) [inline src/api/docker-compose.yml:120]: You shouldn't add the login stuff in this PR, you can remove lines 120-142

File: src/api/env.development
Comment (humphd) [inline src/api/env.development:91]: Remove this extra blank line

File: src/api/posts/Dockerfile
Comment (humphd) [inline src/api/posts/Dockerfile:9]: Change this to:

```
RUN npm install --only=production --no-package-lock
```

File: src/api/posts/package.json
Comment (humphd) [inline src/api/posts/package.json:28]: This is unused, remove it.

File: src/api/posts/package.json
Comment (humphd) [inline src/api/posts/package.json:20]: Both of these elastic deps are unused, remove them.

File: src/api/posts/package.json
Comment (humphd) [inline src/api/posts/package.json:21]: Change this to `^1.x`

File: src/api/posts/package.json
Comment (humphd) [inline src/api/posts/package.json:22]: Remove this, it's unused.

File: src/api/posts/test/posts.test.js
Comment (humphd) [inline src/api/posts/test/posts.test.js:75]: Newline above this

**Review by humphd — CHANGES_REQUESTED** [review]

**Pull request merged**
